lib/puppet/resource.rb

```diff
@@ -71,23 +80,54 @@ def to_data_hash
 
     data["exported"] ||= false
 
-    params = self.to_hash.inject({}) do |hash, ary|
-      param, value = ary
-
+    ext_params = {}
+    params = {}
+    self.to_hash.each_pair do |param, value|
       # Don't duplicate the title as the namevar
-      next hash if param == namevar and value == title
-
-      hash[param] = Puppet::Resource.value_to_pson_data(value)
-      hash
+      unless param == namevar and value == title
+        value = Puppet::Resource.value_to_pson_data(value)
```

Comment:
A test for `is_json_type(value)` on a value that is a `Resource` would return `false`, so yes, this has to be done first to avoid that a `Resource` is serialized as an `ext_parameter`.